### PR TITLE
[MM-34169] Data retention cypress tests

### PR DESCRIPTION
**Pull request by BenCookie95**
#### Summary
Cypress tests for the new data retention feature

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-34169

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: e2e/cypress/integration/enterprise/system_console/compliance/data_retention_policies_spec.js [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+
+                cy.get('#custom_policy_table .DataGrid').within(() => {
+                    cy.uiVerifyCustomPolicyRow(policyId, 'Policy 2', '2 years', '0 teams, 1 channel');
+                });
+
+                cy.apiGetCustomRetentionPolicy(policyId).then((result) => {
+                    expect(result.body.team_count).to.equal(0);
+                    expect(result.body.channel_count).to.equal(1);
+                    expect(result.body.post_duration).to.equal(730);
+                    expect(result.body.display_name).to.equal('Policy 2');
```
Comment (hahmadia) [inline e2e/cypress/integration/enterprise/system_console/compliance/data_retention_policies_spec.js:144]: Don't forget assertion comments and stuff
```
// - [#] indicates a test step (e.g. # Go to a page)
// - [*] indicates an assertion (e.g. * Check the title)
// - Use element ID when selecting an element. Create one if none.
```

**Review by hahmadia — CHANGES_REQUESTED** [review]
Looks good, but lets have more commenting. 
```
// - [#] indicates a test step (e.g. # Go to a page)
// - [*] indicates an assertion (e.g. * Check the title)
// - Use element ID when selecting an element. Create one if none.
```

Assertion comments and normal comments need to be separated by `# / *`

File: e2e/cypress/integration/enterprise/system_console/compliance/data_retention_policies_spec.js (hunk 1 above)
Comment (BenCookie95) [reply e2e/cypress/integration/enterprise/system_console/compliance/data_retention_policies_spec.js]: I added a bunch of comments, let me know if they are ok

File: components/admin_console/data_retention_settings/data_retention_settings.tsx [hunk 2]
```diff
@@ -226,17 +235,37 @@ export default class DataRetentionSettings extends React.PureComponent<Props, St
...
-                    channel_messages: this.getMessageRetentionSetting(policy.post_duration !== -1, policy.post_duration),
-                    applied_to: this.getChannelAndTeamCounts(policy),
+                    description: (
+                        <div id={desciptionId}>
+                            {policy.display_name}
+                        </div>
+                    ),
+                    channel_messages: (
+                        <div id={durationId}>
+                            {this.getMessageRetentionSetting(policy.post_duration !== -1, policy.post_duration)}
```
Comment (hahmadia) [inline components/admin_console/data_retention_settings/data_retention_settings.tsx:256]: We could just pass the post_duration as a param and do the post_duration not equal to -1 within the function

**Review by hahmadia — APPROVED** [review]
One small suggestion

**Comment by saturninoabril** [comment]
@BenCookie95 Totally missed this and sorry for the delay. Looking at this now.

File: e2e/cypress/integration/enterprise/system_console/compliance/data_retention_policies_spec.js [hunk 3]
```diff
@@ -0,0 +1,559 @@
...
+// See LICENSE.txt for license information.
+
+// ***************************************************************
+// - [#] indicates a test step (e.g. # Go to a page)
+// - [*] indicates an assertion (e.g. * Check the title)
+// - Use element ID when selecting an element. Create one if none.
+// ***************************************************************
+
+// Stage: @prod
+// Group: @enterprise @system_console
```
[the remark this answers is not in the document]
Comment (saturninoabril) [reply e2e/cypress/integration/enterprise/system_console/compliance/data_retention_policies_spec.js:11]: Tested and passed. Though I got MM-T4018 failed on first run. So, it's something to keep an eye (but no action needed at the moment).
Screen Shot 2021-05-17 at 5 07 31 PM

File: e2e/cypress/support/api_commands.js [hunk 4]
```diff
@@ -387,6 +387,121 @@ Cypress.Commands.add('apiLinkGroupChannel', (groupID, channelID) => {
...
+        headers: {'X-Requested-With': 'XMLHttpRequest'},
+        url: `/api/v4/data_retention/policies/${id}/channels/search`,
+        method: 'POST',
+        timeout: 60000,
+        body: {term},
+    }).then((response) => {
+        expect(response.status).to.equal(200);
+        return cy.wrap(response);
+    });
+});
```
Comment (saturninoabril) [inline e2e/cypress/support/api_commands.js:503]: We're no longer adding new API commands in this file as we're organizing it at `/support/api/`. With that please move to that folder and create a new file as `/support/api/data_retention.js`. Also, please add corresponding types at ``/support/api/data_retention.d.ts``

File: e2e/cypress/support/api_commands.js [hunk 5]
```diff
@@ -387,6 +387,121 @@ Cypress.Commands.add('apiLinkGroupChannel', (groupID, channelID) => {
...
+ * Get all Custom Retention Policies
+ * @param {Integer} page - The page to select
+ * @param {Integer} perPage - The number of groups per page
+ */
+Cypress.Commands.add('apiGetCustomRetentionPolicies', (page = 0, perPage = 100) => {
+    return cy.request({
+        headers: {'X-Requested-With': 'XMLHttpRequest'},
+        url: `/api/v4/data_retention/policies?page=${page}&per_page=${perPage}`,
+        method: 'GET',
+        timeout: 60000,
```
Comment (saturninoabril) [inline e2e/cypress/support/api_commands.js:400]: Please use the timeout constant at `/fixtures/timeouts`. Same with others.

File: e2e/cypress/support/api_commands.js [hunk 6]
```diff
@@ -387,6 +387,121 @@ Cypress.Commands.add('apiLinkGroupChannel', (groupID, channelID) => {
...
+    }).then((response) => {
+        expect(response.status).to.equal(200);
+        return cy.wrap(response);
+    });
+});
+
+/**
+ * Get Custom Retention Policy teams
+ * @param {string} id - The id of the policy
```
Comment (saturninoabril) [inline e2e/cypress/support/api_commands.js:441]: Please add `page` and `perPage`. Same with others.

**Review by saturninoabril — CHANGES_REQUESTED** [review]
Thanks @BenCookie95! This is great! Overall looks good to me, except for some suggestions.

File: e2e/cypress/support/api_commands.js (hunk 4 above)
Comment (BenCookie95) [reply e2e/cypress/support/api_commands.js]: Per our DM conversation, cypress tests are still using the old matermost-redux repo so data retention types are not available. When cypress tests are updated, I can move these into a typed file.

**Review by saturninoabril — APPROVED** [review]
Thanks @BenCookie95 and noted on follow up to types. Looks good to me, except for the last request.

**Review by furqanmlk — APPROVED** [review]
Thanks @BenCookie95 for fixing failed test case

image

**Pull request merged**